content/posts/oauth3/oauth3.md

```diff
@@ -0,0 +1,46 @@
+
+## PKCE
+
+For å bøte på disse svakhetene har det blitt laget et tillegg til OAuth-standarden. Tillegget beskriver teknikken "Proof Key for Code Exchange" som forkortes "PKCE" og uttales "pixie".
+
+Authorization Code flow ser ut som vist i figuren (se [del 1](/blog/posts/2020/09/oauth-del-1.html) for detaljer).
+
+![authorization code flow](/blog/images/auth_code.png) 
+
+PKCE legger på følgende tillegg:
+ - Klienten genererer en tilfeldig verdi som kalles `code_verifier`. Denne brukes til å generere en `code_challenge` vha en forhåndsavtalt metode. Challengen og `code_challenge_method` legges som ekstra parametre på det initielle `/auth`-kallet til id-provideren.
+ - Id-provider svarer som vanlig, men tar vare på `code_challenge` og `code_challenge_method`.
+ - Klienten sender `authorization_code` som vanlig med i `/token`-kallet, men legger i tillegg på den samme `code_verifier` som ble generert i punkt 1.
+ - Id-provider bruker `code_verifier` til å generere en `code_challenge` på samme måte som klienten gjorde. Hvis de to code challengene ikke er like avvises forespørselen.
+
+Angripere som kjenner din `client_id` og `client_secret` vil dermed ikke kunne gjøre token-kall fordi de ikke kjenner verifieren som ble generert og brukt i punkt 1. Selv om de er i stand til å observere både requesten til og responsen fra `/auth`-endepunktet vil de ikke være i stand til å rekonstruere code challengen.
+
+Standarden definerer to ulike metoder å lage code challenges med: `plain` og `S256`. Plain vil si at code_verifier og code_challenge er samme verdi. For S256 lages code_challenge etter følgende oppskrift: 
+
+```javascript
+base64UrlEncode(sha256(ascii(code_verifier)))
+```
+
+Plain challenges har jo begrenset nytteverdi, så klientene er pålagt å bruke S256 med mindre særlige begrensninger i ressurser eller annet gjør at de ikke er i stand til det. Siden sikkerheten er basert på at code verifieren ikke kan brute forces er det viktig at den er tilfeldig nok (standarden foreslår en entropi på minimum 256 bit). For å holde implementasjonen enkel (og fordi entropien er sterk i utgangspunktet) kreves det ikke salting.
```

Comment:
Er det trivielt å bruteforce dette om man bruker plain? Slik jeg forstod det virker det ganske sikkert selv med plain, og serversiden her bør vel kunne klare oppdage og avverge angrep her? Eller misforstår jeg noe?

Reply:
Plain => verifier og challenge = samme verdi. Hvis man lytter på /auth-kallet kan man bare gjenbruke challenge-verdien som verifier i et /token-kall. Hvis man bruker S256 får man bare hashen ifm /auth-kallet, og da må hashen knekkes for å få tak i verifier til /token-kallet.